src/StirMark/stirmark.1
.TH stirmark 1 "98/07/08" "Stirmark, Version 2.0"
.SH NAME
stirmark \- a tool for robustness testing of image watermarking algorithms
.SH SYNOPSIS
.B stirmark
[
.B options
]
[
.B input file
]
[
.B output file
]
.SH DESCRIPTION
.B StirMark 
is a generic tool for simple robustness testing of image
watermarking algorithms and other steganographic techniques. It can be
applied to photographic digital images and it will distort the
watermark of too simplistic marking techniques such that the embedded
watermark or steganographic message cannot any more be detected and
decoded from the result image.
.PP
.B StirMark 
reads and writes arbitrary images in the PGM and PPM file
formats as they are used in Jef Poskanzer's pbmplus Extended Portable
Bitmap Toolkit. These are simple file formats, and the various PBM
toolkits contain converters between these formats and practically any
other commonly used graphics file format. This way, 
.B StirMark 
can be easily used to process any format on any platform.
.PP
.SH OPTIONS
You can specify an input and an output file name in the command
line. If you omit any of those, 
.I stdin 
and 
.I stdout 
are used instead respectively.
.TP
.BI \-i float
The number of pixel distances that the corner of the target
image is allowed to be inside the original image. If the
number is followed by %, then the maximum corner shift
is defined as a percentage of the image width or height,
whichever is smaller. Default is 
.BI \-i 2.0%.
.TP  
.BI \-o float   
The number of pixel distances that the corner of the target
image is allowed to be outside the original image  
(default 0.7). Sample values taken from outside the
original image are extrapolated. You can also specify a
percentage as with 
.B \-i, 
but this is normally not useful
as 
.B \-o 
values much higher than 1 cause mostly useless extrapolations.
.TP
.BI \-d float
The maximum byte value by which any of the RGB values  
of any pixel is allowed to deviate from the original
value (default 1.5).
.TP
.BI \-n int
Specify the maximum distance of a non-zero impulse response
for the Nyquist lowpass interpolation filter (default 6). 
Runtime depends quadratic on this parameter!
.TP
.BI \-s int
Seed value for the random number generator. If you do  
systematic measurements with 
.B StirMark, 
then repeat the
runs with different 
.B \-s 
values (e.g. from 
.BI \-s 1 
to
.BI \-s 100
), to get significant results regarding the probability with
which 
.B StirMark 
breaks your steganographic technique.
.TP
.BI \-R float
The fraction of pixel displacement allowed for any pixel.
During the distorsion process, any pixel can be move randomly 
along both axes.
.TP
.BI \-b float
The number of pixel displacement allowed for the center
of the image. All the pixel are moved using a smooth  
function. Pixels at the corners are not moved while the
pixel in the center of the image is moved the most.
This is similar to a `punch/pinch' effect.
.TP
.BI \-T name    
The test implemented by default into 
.B StirMark 
include:
.in 10
.ti 9 
\-
rotation by a small angle and cropping: -2, -1, -0.5, 0.5, 1
and 2 degrees. This simulates in part
what is obtain when scanning in image: it is impossible
to align it perfectly.
.br
.ti 9
\-
rotation by a small angle followed by cropping and
rescaling to keep the original size of the image.
.br
.ti 9
\-
scaling by factors 0.5, 0.75, 0.9, 1.1, 1.5 and 2.
.br
.ti 9
\-
centered cropping: 1%, 2%, 5%, 10%, 15%, 20%, 25% and 50%
.br
.ti 9
\- 
median filtering
.br
.ti 9
\- 
gaussian filtering (blurring)
.br
.ti 9
\- 
horizontal flip
.br
.ti 9
\- 
original 
.B StirMark 
.br
.ti 9
\-
.B StirMark 
with randomisation using the default
randomisation parameter
.br
.ti 9
\- 
.B StirMark
with randomisation and bending effect using default parameters.
.PP
Normally you do not have to use any of those options (except 
.B \-T
) and
you should try the default values first. If this doesn't remove the
watermark, then the embedding algorithm is not too bad and you can try to
increase the 
.B \-i 
value to get a stronger geometric transform. Then try
higher 
.B \-d 
values to get a more significant color transform or 
.B \-R 
or 
.B \-b 
to increase the random distorsion. If you want no geometric transform at
all, set 
.BI \-i 0 
and 
.BI \-o 0
.
.PP
For example 
.P
.ft R
stirmark -Tbase test.ppm
.ft P
.P
will create 32 files with base name `base' and different suffixes
depending
on the transformation applied.
.P
You can still use the `original` 
.B StirMark:
.P                          
.ft R
stirmark -i0 -o0 -d0 test.ppm output.ppm
.ft P
.P
does not modify the image at all.
.PP
.B StirMark 
basically simulates a resampling process, i.e. it introduces
the same kind of errors into an image that you would expect if you
printed an image on a high-quality printer and then scan the image
again with a high-quality scanner. The algorithm applies a minor
geometric distortion, i.e. the image is slightly stretched, sheared,
shifted, and/or rotated by an unnoticeable random amount and then
resampled using Nyquist interpolation. In addition, a transfer function
that introduces a small and smoothly distributed error into all sample
values is applied, which acts like a small non-linear analog/digital
converter imperfection typically found in scanners and display devices.
.PP
Finally we added a randomisation process by which each pixel is   
slightly shift along th x and y axis. This is also emphasized by a
punch/pinch effect. The difference on drawings with straight lines
is usually obvious however it almost unoticeable with photos. This  
increase the size of the space that a watermark detector should  
search.
.PP
.B StirMark 
introduces a practically unnoticeable quality loss in the  
image if it is applied only once. However after a few iterated
applications of the algorithm, the image degradation becomes soon very
noticeable. This makes StirMark (especially the 
.B \-d 
option) also a 
model for algorithms that can be applied to artificially degrade  
images on replay to get rid of the lossless-copying property of   
digital multimedia objects.
.PP
.B StirMark 
is a generic steganographic attack tool in that it is not
targeted against any specific steganographic algorithm. I hope to
establish with
.B StirMark 
a reasonable minimum robustness requirement 
for steganographic image processing tools, i.e. if information
embedded by an algorithm into an image does not survive the 
.B StirMark  
process, then the steganographic technique used should considered to
be unacceptably easy to break.
.PP
Version 2.0 adds the following command line option to version 1.0
.TP
.BI \-T name 
to perform a standard series of tests on a single image.
.TP
.BI \-R float
and 
.BI \-b float
are used in the enhanced distortion process.
.PP
.SH AUTHOR
Suggestions for improving this software are always welcome. We are   
also interested in robustness test results obtained by processing with
.B StirMark 
images that were marked with various available watermarking tools.
.PP
Markus G. Kuhn and Fabien A. P. Petitcolas
.PP
Markus G. Kuhn and Fabien A. P. Petitcolas are with the Computer
Laboratory, University of Cambridge, New Museum Site, CB2 3QG, UK.
.P
M. G. Kuhn               http://www.cl.cam.ac.uk/~mgk25/
.P
F. A. P. Petitcolas      http://www.cl.cam.ac.uk/~fapp2/   
.P
.SH NOTES
Surprisingly, first experiences with various watermarking tools show
that the watermarks of
.B all
tools we have been able to locate until today (DigiMark, JK_PGS, SysCop,
SureSign, etc.) can be removed or made undetectable by 
.B StirMark,
especially if 
.B StirMark 
is followed by a medium quality JPEG compression. For more
information, please see:
.P
     http://www.cl.cam.ac.uk/~fapp2/steganography/
.P
A paper describing among other things our experiences with 
.B StirMark 
in more detail is available from:
.P   
     http://www.cl.cam.ac.uk/~fapp2/papers/ih98-attacks/
.P
This man-page has been created by Peter Meerwald, pmeerw@cosy.sbg.ac.at
from the README, Version 1.2, 1998/07/07 by F. A. P. Petitcolas.
.SH AVAILABILITY
.B StirMark 
is freely available under the GNU General Public Licence in
full portable ISO C source code from
.P
     http://www.cl.cam.ac.uk/~fapp2/watermarking/stirmark/
.P
The original 
.B StirMark 
(version 1.0) is still available from   
.P
     http://www.cl.cam.ac.uk/~mgk25/stirmark/
.P
PBM tools are for instance available from
.P
     ftp://ftp.wustl.edu/graphics/graphics/packages/pbmplus/pbmplus10dec91.tar.Z
     ftp://gatekeeper.dec.com/.b/X11-contrib/pbmplus10dec91.tar.Z
     ftp://oak.oakland.edu/pub/msdos/graphics/pbmp191d.zip
.P
but you'll need these tools only if you do not already have other
image processing software that can handle the PPM or PGM file formats.


src/StirMark/stirmark.c
/*----------------------------------------------------------------------------
 * StirMark -- Experimental watermark resilience benchmark
 *
 * Markus G. Kuhn <mkuhn@acm.org>, University of Cambridge
 * Fabien A. P. Petitcolas <fapp2@cl.cam.ac.uk>, University of Cambridge
 *
 * History: Version 1.0, 1997-11-10 by Markus G. Kuhn
 *             This is the initial version of StirMark tuned
 *             after various experiments on some existing
 *             steganographic algorithms.
 *
 *          Version 1.1, 1998-01-30 by Fabien A. P. Petitcolas
 *             Inverse StirMark added: anew option allows you
 *             to save the parameters used during the distortion
 *             and to inverse the process later on.
 *
 *          Version 1.2, 1998-03-11 by Fabien A. P. Petitcolas
 *             Add a ramdomise option: the interpolation process
 *             is randomised by changing slightly the coordinates
 *             of the points using the _randomise() fuction.
 *
 *          Version 1.3, 1998-05-19 by Fabien A. P. Petitcolas
 *             Add a global bending (x and y) to the picture.
 *             In fact a slight punch effect.
 *
 *          Version 2.0, 1998-06-15 by Fabien A. P. Petitcolas
 *             Add the simple transformation functions and
 *             filtering routines for the `benchmark'.
 *
 *          Version 2.1, 1998-08-15 by Fabien A. P. Petitcolas
 *             Add JPEG support using the library of the
 *             Independent JPEG Group.
 *
 *          Version 2.2, 1998-09-17 by Fabien A. P. Petitcolas
 *             Minor corrections and bug fixes.
 *
 *          Version 2.3, 1998-19-10 by Fabien A. P. Petitcolas
 *             Improved the speed of StirMark by changing the
 *             resampling method. Now use code by Neil Dodgson,
 *             Graphics Group, University of Cambridge. The
 *             speed has been increased by a factor 8 to 9.
 *             Most of the old StirMark code has been re-organised
 *             and clarified.
 *
 *          Version 3.0, 1999-20-01 by Fabien A. P. Petitcolas
 *             New transformations: row and columns removal,
 *             shrearing in X and Y direction and sharpening.
 *             Color quantisation support based on Jef Poskanzer's
 *             code. Laplacian based removal attack using Richard
 *             Barnett's code. Quality measure functions: SNR and
 *             PSNR. JPEG support for input image. Autodection of
 *             the input format based on file extension (pgm, ppm,
 *             jpg or jpeg).
 *
 *          Version 3.1, 1999-04-07 by Fabien A. P. Petitcolas
 *             Minor improvement: -NOJPEG prevent JPEG compression
 *             after geometrical distortions.
 *             Possibility to select a category of tests only.
 *             New tests: general linear geometric distortions, change
 *             of aspect ration, different sizes of Median filtering,
 *             assymetric row and column removal,
 *             new default values for some existing tests.
 *             Option to save and reuse parameters of geometrical 
 *             distortions.
 *
 *          Version 3.2,
 *             Corrected sizes of Median filters
 *             /? /h -? -h
 *
 * StirMark applies a number of minor almost invisible distortions to
 * an image in order to test whether some watermark detector can still
 * find the embedded signal after this processing. The distortion simulate
 * a resampling process, i.e. they include a minor geometric transform,
 * linear interpolation, slightly non-identical transfer functions as
 * well as highly non-linear and non bi-linear geometrical distortions.
 * The distortions are adjusted to be not recognizeable for typical 
 * photographic source images by human viewers.
 *
 * Copyright (c) 1997, 1999 Fabien A. P. Petitcolas and Markus G. Kuhn
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any non-commercial purpose and without fee is hereby
 * granted (GPL), provided that the above copyright notice appear in all
 * copies and that both that copyright notice and this permission notice
 * appear in supporting documentation. This software is provided "as is" 
 * without express or implied warranty. The authors shall not be held
 * liable in any event for incidental or consequential damages in
 * connection with, or arising out of, the furnishing, performance, or
 * use of this program.
 * 
 * If you use StirMark for your research, please cite:
 *
 *   Fabien A. P. Petitcolas, Ross J. Anderson, Markus G. Kuhn. Attacks
 *   on copyright marking systems, in David Aucsmith (Ed), Information
 *   Hiding, Second International Workshop, IH'98, Portland, Oregon,
 *   USA, April 15--17, 1998, Proceedings, LNCS 1525, Springer-Verlag,
 *   ISBN 3-540-65386-4, pp. 219--239.
 *
 *   <http://www.cl.cam.ac.uk/~fapp2/papers/ih98-attacks/>
 *
 * and 
 *
 *   Martin Kutter and Fabien A. P. Petitcolas. A fair benchmark for
 *   image watermarking systems, To in E. Delp et al. (Eds), in
 *   vol. 3657, proceedings of Electronic Imaging '99, Security and
 *   Watermarking of Multimedia Contents, San Jose, CA, USA, 25--27
 *   January 1999. The International Society for Optical
 *   Engineering. To appear.
 *
 *   <http://www.cl.cam.ac.uk/~fapp2/papers/ei99-benchmark/>
 *
 * See the also the "Copyright" file provided in this package for
 * copyright information about code and libraries used in StirMark.
 *
 * $Header: /StirMark/stirmark.c 22    11/08/99 19:28 Fapp2 $
 *----------------------------------------------------------------------------
 */

#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <time.h>

#include "error.h"
#include "bench.h"
#include "stirmark.h"
#include "image.h"
#include "quality.h"

#ifndef min
#define min(a,b)    (((a) < (b)) ? (a) : (b))
#endif

extern int cutoff;

/* From bench.c */
extern void StirMark(IMAGE I, OPTIONS o, NYQUIST n, FILE *stream, int use_jpeg);
extern void Benchmark(IMAGE I, NYQUIST n, char *strzBasename, int nTestSet);
extern int  LRAttack(IMAGE *pSImage, 
                     IMAGE *pDImage, 
                     double dStrength, 
                     double dJNDStrength, 
                     int nScale);

/*----------------------------------------------------------------------------
 */
static FILE *StreamOpen(const char *strzFileName, const char *strzMode)
{
    FILE *pStream;

    if ((strzFileName == NULL) || (strzMode == NULL))
        ERROR("StreamOpen: null parameter");

    pStream = fopen(strzFileName, strzMode);
    if (pStream == NULL)
        ERROR("StreamOpen: cannot open file '%s' (mode %s)",
              strzFileName, strzMode);

    return (pStream);
}

/*----------------------------------------------------------------------------
 */
void Usage(char *name)
{
    char usage[] =
        "StirMark %s\nUsage: %s [options] [<input file> [<output file>]]\n"
        "\t-h, -?, /h, /?\tfor this help message\n\n"
        "\t-b<float>\tbending factor (%.2f)\n"
        "\t-d<float>\tmaximum variation of a pixel value (%.2f)\n"
        "\t-i<float>\tmaximum distance a corner can move inwards (%.2f%%)\n"
        "\t-ll<n><float>\tLR strengths for scale 2^n\n"
        "\t-lj<n><float>\tJND LR strengths for scale 2^n\n"
        /*"\t-n<int>\t\tuse Nyquist interpolation up to specified distance (%d)\n"*/
        "\t-o<float>\tmaximum distance a corner can move outwards (%.2f%%)\n"
        "\t-PS<filename>\tsave main distortion parameters into file\n"
        "\t-PL<filename>\tload main distortion parameters from file\n"
        "\t-PS+<filename>\tsave all distortion parameters into file\n"
        "\t-PL+<filename>\tload all distortion parameters from file\n"
        "\t\t\tSame format as input\n"
        "\t-q<int>\t\tJPEG quality factor of output(%d)\n"
        "\t-NOJPEG\t\tno JPEG compression after geometric distortions\n"
        "\t-R<float>\trandomisation factor (%.2f)\n"
        "\t-s<int>\t\trandom number generator seed\n"
        "\t-t<float>\trelative modification of Nyquist filter threshold (%.2f)\n"
        "\n"
        "\t-PSNR<original>\tcompute the PSNR of the input file (original image required)\n"
        "\n"
        "\t-T<name>\tbenchmark: series of pre-defined simple transformations\n"
        "\t\t\tand distortions. All other options ignored expect: \n"
        "\t-S<int>\t\tset of transformations used for test (requires -T option):\n"
        "\t\t\tconvolution filters (%d), median filter (%d), FMLR (%d),\n"
        "\t\t\tJPEG compression (%d), colour quantisation (%d), scaling (%d),\n"
        "\t\t\tshearing (%d), aspect ratio (%d), general linear (%d), \n"
        "\t\t\trotation crop (%d), rotation crop scale (%d), cropping(%d), \n"
        "\t\t\tflip (%d), row and col removal (%d), geometric distortions (%d)\n\n";
    
    if (name == NULL) exit(1);    

    fprintf(stderr, usage, VERSION, name, BENDING_FACTOR, DEVIATION,
        INWARDS, OUTWARDS, JPEG_QUALITY, RANDOM_FACTOR, ROLLOFF,
        TEST_FILTERING, TEST_MEDIAN_FILTERING, TEST_FMLR,
        TEST_COMPRESSION, TEST_QUANTISE, TEST_SCALING,
        TEST_SHEAR, TEST_ASPECT_RATIO, TEST_GENERAL_LINEAR,
        TEST_ROTATION_CROP, TEST_ROTATION_SCALE, TEST_CROPPING,
        TEST_FLIP, TEST_ROW_COL_REMOVAL, TEST_GEOM_DISTORTIONS);

    exit(1);
}


/*----------------------------------------------------------------------------
 */
int main(int argc, char **argv)
{
    char basename[MAX_BASENAME_LENGTH + 1];
    char *pfin = NULL, *pfout = NULL, *pfpsnr = NULL;
    FILE *fin = stdin, *fout = stdout;
    double do_lr[6];    /* LR strengths */
    double do_jndlr[6]; /* JND LR strengths */
    double psnr;        /* Value of PSRN is computed */
    int do_test = 0;    /* Whether -T option is used */
    int use_jpeg = 1;   /* For -NOJPEG option */
    int nTestSet = TEST_ALL; /* For -S option */
    int i = 0, j = 0;
    IMAGE   I, oI;
    OPTIONS o;
    NYQUIST n;
    
    for (i = 0; i < 6; i++)
    {
        do_lr[i] = 0;     /* clear strength arrays */
        do_jndlr[i] = 0;
    }            

    /* Default parameters */
    o.inwards = INWARDS;
    o.outwards = OUTWARDS;
    o.relative_inwards = RELATIVE_INWARDS;
    o.relative_outwards = RELATIVE_OUTWARDS;
    o.random_factor = RANDOM_FACTOR;
    o.bending_factor = BENDING_FACTOR;
    o.deviation = DEVIATION;
    o.jpeg_quality = JPEG_QUALITY_STIR;
    o.parameters = DIST_OP_NEW;

    n.rolloff.x = n.rolloff.y = ROLLOFF;
    n.highpass.x = n.highpass.y = HIGHPASS;
    n.cutoff = CUTOFF;

    /* Seed the random-number generator with current time so that */
    /* the numbers will be different every time we run.           */
    srand((unsigned)time(NULL));

    /* Read command line arguments */
    for (i = 1; i < argc; i++)
    {
        if ((argv[i][0] == '/' || argv[i][0] == '-') && 
            (argv[i][1] == '?' || argv[i][1] == 'h') &&
             argv[i][2] == 0)

            Usage(argv[0]);
        else if (argv[i][0] == '-')
        {
            for (j = 1; j > 0 && argv[i][j] != 0; j++)
            {
                switch (argv[i][j])
                {
                case 'i': /* inward parameter */
                    o.inwards = atof(argv[i] + j + 1);
                    o.relative_inwards = argv[i][strlen(argv[i])-1] == '%';
                    j = -1;
                    break;
                case 'o': /* outward parameter */
                    o.outwards = atof(argv[i] + j + 1);
                    o.relative_outwards = argv[i][strlen(argv[i])-1] == '%';
                    j = -1;
                    break;
                case 'd': /* luminance modulation */
                    o.deviation = atof(argv[i] + j + 1);
                    j = -1;
                    break;
                case 't': /* Nyquist filter threshold */
                    if (argv[i][j+1] == 'x')
                        n.rolloff.x = atof(argv[i] + j + 2);
                    else if (argv[i][j+1] == 'y')
                        n.rolloff.y = atof(argv[i] + j + 2);
                    else
                        n.rolloff.x = n.rolloff.y = atof(argv[i] + j + 1);
                    j = -1;
                    break;
                case 'R': /* randomisation factor */
                    o.random_factor = atof(argv[i] + j + 1);
                    j = -1;
                    break;
                case 'b': /* bending factor */
                    o.bending_factor =  atof(argv[i] + j + 1);
                    j = -1;
                    break;
                case 's': /* Seed */
                    srand(atoi(argv[i] + j + 1));
                    j = -1;
                    break;
                case 'S': /* Test set */
                    nTestSet = atoi(argv[i] + j + 1);
                    j = -1;
                    break;
                /*case 'n': * Nyquist distance *
                    * use true Nyquist sin(x)/x interpolation *
                    if (isdigit(argv[i][j+1]))
                    {
                        n.cutoff = atoi(argv[i] + j + 1);
                        j = -1;
                    }
                    break;*/
                case 'q':
                    o.jpeg_quality = atoi(argv[i] + j + 1);
                    j = -1;
                    break;
                case 'P':
                    /* Compute PSNR */
                    if (strncmp(argv[i] + j, "PSNR", 4) == 0)
                        pfpsnr = argv[i] + j + 4;

                    /* Save or load distortion parameters*/
                    else if (strncmp(argv[i] + j, "PS+", 3) == 0)
                    {
                        o.pszFileName = argv[i] + j + 3;
                        o.parameters = DIST_OP_SAVE_ALL;
                    }
                    else if (strncmp(argv[i] + j, "PL+", 3) == 0)
                    {
                        o.pszFileName = argv[i] + j + 3;
                        o.parameters = DIST_OP_LOAD_ALL;
                    }
                    else if (strncmp(argv[i] + j, "PS", 2) == 0)
                    {
                        o.pszFileName = argv[i] + j + 2;
                        o.parameters = DIST_OP_SAVE;
                    }
                    else if (strncmp(argv[i] + j, "PL", 2) == 0)
                    {
                        o.pszFileName = argv[i] + j + 2;
                        o.parameters = DIST_OP_LOAD;
                    }
                    else
                        ERROR("Unrecognised option -P??.");
                    j = -1;
                    break;
                case 'N':
                    /* PPM ouput when not in -T mode */
                    if (strncmp(argv[i] + j, "NOJPEG", 6) == 0)
                        use_jpeg = 0;
                    else
                        ERROR("Unrecognised option -N??.");
                    j = -1;
                    break;
                case 'l':
                    /* Parameters for the FMLR attack */
                    {
                    switch (argv[i][j+1])
                    {
                    case 'l':
                    {
                        switch (argv[i][j+2])
                        {
                        case '1': /* Scale 1 */
                            do_lr[0] = atof(argv[i] + j + 3);
                            break;
                        case '2': /* Scale 2 */
                            do_lr[1] = atof(argv[i] + j + 3);
                            break;
                        case '3': /* Scale 4 */
                            do_lr[2] = atof(argv[i] + j + 3);
                            break;
                        case '4': /* Scale 8 */
                            do_lr[3] = atof(argv[i] + j + 3);
                            break;
                        case '5': /* Scale 16 */
                            do_lr[4] = atof(argv[i] + j + 3);
                            break;
                        case '6': /* Scale 32 */
                            do_lr[5] = atof(argv[i] + j + 3);
                            break;                
                        default:
                            ERROR("Scale not supported for -ll option.");
                        }
                        j = -1;
                        break;
                    }
                    case 'j':
                    {
                        switch (argv[i][j+2])
                        {
                        case '1': /* Scale 1 */
                            do_jndlr[0] = atof(argv[i] + j + 3);
                            break;
                        case '2': /* Scale 2 */
                            do_jndlr[1] = atof(argv[i] + j + 3);
                            break;
                        case '3': /* Scale 4 */
                            do_jndlr[2] = atof(argv[i] + j + 3);

                            break;
                        case '4': /* Scale 8 */
                            do_jndlr[3] = atof(argv[i] + j + 3);
                            break;
                        case '5': /* Scale 16 */
                            do_jndlr[4] = atof(argv[i] + j + 3);
                            break;
                        case '6': /* Scale 32 */
                            do_jndlr[5] = atof(argv[i] + j + 3);
                            break;
                        default:
                            ERROR("Scale not supported for -lj option.");
                        }
                        j = -1;
                        break;
                    }
                    default:
                        ERROR("Unrecognised option -l??.");
                    }
                    break;
                    }
                case 'T':
                    do_test = 1;
                    strncpy(basename, argv[i] + j + 1,
                            min(strlen(argv[i] + j + 1), MAX_BASENAME_LENGTH));
                    basename[min(strlen(argv[i] + j + 1),
                             MAX_BASENAME_LENGTH)] = '\0';
                    if (strlen(basename) == 0)
                        ERROR("Empty base name.");
                    j = -1;
                    break;
                default:
                    ERROR("One of the options has not been recognised. "
                          "Try -? for help.");
                }
            }
        }
        else if (pfin == NULL)
            pfin = argv[i];
        else if (pfout == NULL)
            pfout = argv[i];
        else
            ERROR("One of the options has not been recognised. "
                  "Try -? for help.");
    }
    if ((nTestSet != TEST_ALL) && (do_test == 0))
        ERROR("The -S option must be used with -T option.");

    if (n.rolloff.x < 0) n.rolloff.x = -n.rolloff.x, n.highpass.x = 1;
    if (n.rolloff.y < 0) n.rolloff.y = -n.rolloff.y, n.highpass.y = 1;

    if (pfin)
        ImageRead(&I, pfin);
    else
        ImageReadPPM(&I, fin);
    if (pfout)
        fout = StreamOpen(pfout, "wb");

    if (pfpsnr)
    {
        /* Just compute sthe PSNR and exit */
        if (pfin == NULL)
            printf("Original: %s - Modified: stdin - ", pfpsnr);
        else
            printf("Original: %s - Modified: %s - ", pfpsnr, pfin);
        ImageRead(&oI, pfpsnr);
        if ((psnr = PSNR(oI, I)) <= MIN_QUALITY)
            printf("PSNR not meaningful\n");
        else
            printf("PSNR = %f\n", psnr);
        ImageClear(&oI);
    }
    else
    {
        if (do_test)
            Benchmark(I, n, basename, nTestSet);
        else /* Backward compatibility */
        {
            /* First the LR attack */
            for (i = 0; i < 6; i++)
                /* Nothing is done if stregths are 0 */
                LRAttack(&I, &I, do_lr[i], do_jndlr[i], (1 << i));

            /* Then the StZirMark geometrical distortions */
            StirMark(I, o, n, fout, use_jpeg);
        }
    }

    if (pfout)
        fclose(fout);

    ImageClear(&I);
    return (0);
}

